starknet_py/proxy/contract_abi_resolver.py

```diff
@@ -0,0 +1,155 @@
+    """
+    Proxy resolving configuration
+    """
+
+    proxy_checks: List[ProxyCheck]
+    """
+    List of classes implementing :class:`starknet_py.proxy.proxy_check.ProxyCheck` ABC,
+    that will be used for checking if contract at the address is a proxy contract.
+    """
+
+
+def prepare_proxy_config(proxy_config: ProxyConfig) -> ProxyConfig:
+    if "max_steps" in proxy_config:
+        warnings.warn(
+            "ProxyConfig.max_steps is deprecated. Contract.from_address always makes at most 1 step.",
+            category=DeprecationWarning,
+        )
+    proxy_checks = [OpenZeppelinProxyCheck(), ArgentProxyCheck()]
+    return {"proxy_checks": proxy_checks, **proxy_config}
+
+
+class ImplementationType(Enum):
+    """
+    Enum representing transaction types
+    """
+
+    CLASS_HASH = "class_hash"
+    ADDRESS = "address"
+
+
+class ContractAbiResolver:
+    """
+    Class for resolving abi of a contract
+    """
+
+    def __init__(
+        self,
+        address: Address,
+        client: Client,
+        proxy_config: ProxyConfig,
+    ):
+        """
+        :param address: Contract's address
+        :param client: Client used for resolving abi
+        :param proxy_config: Proxy config for resolving proxy
+        """
+        self.address = address
+        self.client = client
+        self.proxy_config = proxy_config
+
+    async def resolve(self) -> Abi:
+        """
+        Returns abi of either direct contract or contract proxied by direct contract depending on proxy_config.
+        :raises ContractNotFoundError: when contract could not be found at address
+        :raises ProxyResolutionError: when given ProxyChecks were not sufficient to resolve proxy
+        :raises AbiNotFoundError: when abi is not present in contract class at address
+        """
+        if self.proxy_config:
+            return await self.resolve_abi()
+        return await self.get_abi_for_address()
+
+    async def get_abi_for_address(self) -> Abi:
+        """
+        Returns abi of a contract directly from address.
+        :raises ContractNotFoundError: when contract could not be found at address
+        :raises AbiNotFoundError: when abi is not present in contract class at address
+        """
+        contract_class = await self._get_class_by_address(address=self.address)
+        if contract_class.abi is None:
+            raise AbiNotFoundError()
+        return contract_class.abi
+
+    async def resolve_abi(self) -> Abi:
+        """
+        Returns abi of a contract that is being proxied by contract at address.
+        :raises ContractNotFoundError: when contract could not be found at address
+        :raises ProxyResolutionError: when given ProxyChecks were not sufficient to resolve proxy
+        :raises AbiNotFoundError: when abi is not present in proxied contract class at address
+        """
+        impl = await self._get_implementation_from_proxy()
+        implementation, implementation_type = impl
```

Comment:
Can we combine those lines?

Reply:
No, because I'd really like to keep the full names of the variables and in one line case this abomination is output when running `poe format_diff`:
```
-        implementation, implementation_type = await self._get_implementation_from_proxy()
+        (
+            implementation,
+            implementation_type,
+        ) = await self._get_implementation_from_proxy()
```

Reply:
Okay, let's keep it that way then.